File: lib/ai/adherenceAgent.ts
export interface AdherenceDecision {
  acknowledgement_count: number;
  previous_count?: number;
  adherence_status: string;
  escalation_level: 'none' | 'gentle' | 'caregiver_alert' | 'doctor_escalation';
  action: string;
  reason: string;
  requires_caregiver_alert?: boolean;
  requires_doctor_alert?: boolean;
}

export class AdherenceEscalationAgent {
  private counter: number = 0;
  private maxCounter: number = 5;

  public resetCounter() {
    this.counter = 0;
  }

  public getCounter(): number {
    return this.counter;
  }

  public recordAcknowledgment(_doseId: string = 'current_dose', taken: boolean = true): AdherenceDecision {
    const prev = this.counter;
    if (taken) this.counter = 0;

    return {
      acknowledgement_count: this.counter,
      previous_count: prev,
      adherence_status: 'on_track',
      escalation_level: 'none',
      action: 'log_success',
      reason: 'Dose successfully marked as taken by patient.'
    };
  }

  public recordMissedDoseOrUnanswered(): AdherenceDecision {
    this.counter = Math.min(this.maxCounter, this.counter + 1);
    const c = this.counter;

    if (c === 1) {
      return {
        acknowledgement_count: 1,
        adherence_status: 'day_1_missed',
        escalation_level: 'gentle',
        action: 'send_patient_reminder',
        reason: 'Day 1: Medication reminder unanswered. Gentle in-app prompt and Telegram reminder sent to patient.',
        requires_caregiver_alert: false,
        requires_doctor_alert: false
      };
    } else if (c === 2) {
      return {
        acknowledgement_count: 2,
        adherence_status: 'day_2_missed',
        escalation_level: 'gentle',
        action: 'send_patient_reminder',
        reason: 'Day 2: Second consecutive missed dose. Gentle reassurance and safety guideline sent to patient.',
        requires_caregiver_alert: false,
        requires_doctor_alert: false
      };
    } else if (c === 3 || c === 4) {
      return {
        acknowledgement_count: c,
        adherence_status: `day_${c}_caregiver_alert`,
        escalation_level: 'caregiver_alert',
        action: 'dispatch_caregiver_alert',
        reason: `Day ${c}: Patient has not taken medication for ${c} consecutive days. Urgent Telegram alert dispatched to Caregiver Priya to check in on patient.`,
        requires_caregiver_alert: true,
        requires_doctor_alert: false
      };
    } else {
      return {
        acknowledgement_count: 5,
        adherence_status: 'day_5_doctor_escalation',
        escalation_level: 'doctor_escalation',
        action: 'dispatch_doctor_escalation',
        reason: 'Day 5: Patient has missed medication for 5 consecutive days. Critical clinical escalation dispatched to Dr. Aarav Mehta (Physician) and Apollo Clinical Hub.',
        requires_caregiver_alert: true,
        requires_doctor_alert: true
      };
    }
  }
}

export const adherenceEscalationAgent = new AdherenceEscalationAgent();

File: ai/adherence_agent.py
"""
Agent 2 — Adherence Escalation Agent.
Role: Patient Monitoring Manager & Adherence Decision Engine.
Responsibilities:
- Tracks medication reminder acknowledgments and timestamps.
- Maintains the 1–5 non-acknowledgment counter.
- Executes configurable escalation policy tree:
  Counter 1 -> Gentle reminder
  Counter 2 -> Stronger / clear reminder
  Counter 3 -> Flag adherence concern & Caregiver consideration
  Counter 4 -> Direct Caregiver WhatsApp / SMS alert
  Counter 5 -> Clinical / Physician escalation
- Returns structured adherence decision state.
"""

from typing import Dict, Any, Optional

class AdherenceEscalationAgent:
    def __init__(self):
        self.non_acknowledgment_counter = 0
        self.max_counter = 5
        self.adherence_history = []

    def record_acknowledgment(self, dose_id: str, taken: bool = True) -> Dict[str, Any]:
        """Resets the non-acknowledgment counter upon patient acknowledgment or intake."""
        prev_count = self.non_acknowledgment_counter
        if taken:
            self.non_acknowledgment_counter = 0
        return {
            "event": "DOSE_ACKNOWLEDGED" if taken else "DOSE_POSTPONED",
            "acknowledgement_count": self.non_acknowledgment_counter,
            "previous_count": prev_count,
            "adherence_status": "on_track",
            "escalation_level": "none",
            "action": "log_success",
            "reason": "Dose successfully marked as taken by patient."
        }

    def record_missed_dose_or_unanswered(self, dose_id: Optional[str] = None) -> Dict[str, Any]:
        """Increments non-acknowledgment counter and computes escalation tree decision."""
        self.non_acknowledgment_counter = min(self.max_counter, self.non_acknowledgment_counter + 1)
        count = self.non_acknowledgment_counter

        if count == 1:
            return {
                "acknowledgement_count": count,
                "adherence_status": "mild_delay",
                "escalation_level": "gentle",
                "action": "send_gentle_reminder",
                "reason": "1st reminder unanswered. Gentle follow-up prompt required.",
                "requires_caregiver_alert": False,
                "requires_doctor_alert": False
            }
        elif count == 2:
            return {
                "acknowledgement_count": count,
                "adherence_status": "delayed",
                "escalation_level": "reminder",
                "action": "send_clear_reminder",
                "reason": "2nd reminder unanswered. Clear single-action reminder required.",
                "requires_caregiver_alert": False,
                "requires_doctor_alert": False
            }
        elif count == 3:
            return {
                "acknowledgement_count": count,
                "adherence_status": "at_risk",
                "escalation_level": "caregiver_consideration",
                "action": "prepare_caregiver_notice",
                "reason": "3 consecutive missed prompts. Adherence risk detected.",
                "requires_caregiver_alert": False,
                "requires_doctor_alert": False
            }
        elif count == 4:
            return {
                "acknowledgement_count": count,
                "adherence_status": "high_risk",
                "escalation_level": "caregiver",
                "action": "contact_caregiver",
                "reason": "4 consecutive missed prompts. Dispatching verified WhatsApp alert to Priya.",
                "requires_caregiver_alert": True,
                "requires_doctor_alert": False
            }
        else: # count >= 5
            return {
                "acknowledgement_count": count,
                "adherence_status": "critical",
                "escalation_level": "doctor",
                "action": "contact_doctor_and_caregiver",
                "reason": "5 consecutive missed reminders. Escalating to Dr. Mehta and Caregiver Priya for clinical review.",
                "requires_caregiver_alert": True,
                "requires_doctor_alert": True
            }

    def get_current_status(self) -> Dict[str, Any]:
        """Returns the current state of the adherence escalation counter."""
        count = self.non_acknowledgment_counter
        status = "on_track" if count == 0 else "at_risk" if count < 4 else "critical"
        return {
            "acknowledgement_count": count,
            "adherence_status": status,
            "max_counter": self.max_counter
        }

    def reset_counter(self):
        self.non_acknowledgment_counter = 0
